### DISCUSSION TITLE
atomWithStorage with Capacitorjs

### DISCUSSION THREAD

**Discussion by rrdesignweb** [pmndrs/jotai#1642] [Q&A]
Hi there wasn't sure what category to add this in (new feature or q&a etc)

I am looking to use my current nextjs app with capacitorjs - https://capacitorjs.com/.  I have atomWithStorage currently implemented.

Wanted to know how I would go about implementing localStorage - https://capacitorjs.com/docs/apis/preferences states "_Mobile OSs may periodically clear data set in window.localStorage, so this API should be used instead. This API will fall back to using localStorage when running as a Progressive Web App._"

Thanks in advance

**Comment by dai-shi** [comment]
`atomWithStorage` accepts custom `storage` option. So, it should allow integration with `@capacitor/preferences`.

https://github.com/pmndrs/jotai/blob/3b7be881b483c966179c7a2f15d76839d1ef5b66/src/utils/atomWithStorage.ts#L120

**Comment by rrdesignweb** [comment]
Hi there, thank you, I have got it working ok, but having some small issues, if you are able to assist (I am probably missing something basic)

I have an app where I am using capacitor and  googleAuth (https://github.com/CodetrixStudio/CapacitorGoogleAuth)
I retrieve email, and name etc from googleAuth details and store it in atomWithStorage

I login to the app with routes me to dashboard where I display email and name which works well, however when I refresh the page it doesn't seem to get parsed properly ends up looking like `{value: '{"email":"....","givenName":"...","familyName":"..."}'}` and displays as `undefined` in the UI, not sure where I am going wrong?

<pre>
const customStorage = {
    setItem: (key: string, value: any) => {
        const val = JSON.stringify(value)
        Preferences.set({
            key,
            value: val
        });
    },
    getItem: (key: string) => {
        return Preferences.get({ key }); // <--?
    },
    removeItem: (key: string) => {
        Preferences.remove({ key });
    },
};

export const userSessionAtom = atomWithStorage(
    'googleUser',
    "",
    customStorage
);

//Google Auth Login
const [, setUserSession] = useAtom(userSessionAtom)
const [error, setError] = useState(null);
...
const login = async () => {
        try {
            const { email, givenName, familyName } = await GoogleAuth.signIn();
            setUserSession({ email, givenName, familyName });
        } catch (e) {
            setError(e);
        }
    }
    ...
</pre>

References - https://capacitorjs.com/docs/apis/preferences#example

Thanks in advance!

**Reply by dai-shi** [reply]
Maybe this:
```diff
      getItem: (key: string) => {
-         return Preferences.get({ key }); // <--?
+         return JSON.parse(Preferences.get({ key }).value);
      },
```

**Comment by kavdev** [comment]
`preferences.store.ts`
```typescript
import { Preferences } from "@capacitor/preferences"
import { atomWithStorage, createJSONStorage } from "jotai/utils"
import { AsyncStringStorage } from "jotai/vanilla/utils/atomWithStorage"

const preferencesStorage: AsyncStringStorage = {
    getItem: async (key: string) => {
        const { value } = await Preferences.get({ key })
        return value
    },
    setItem: (key: string, newValue: string) => Preferences.set({ key, value: newValue }),
    removeItem: (key: string) => Preferences.remove({ key }),
}

export const sessionPreferenceAtom = atomWithStorage<string>(
    "session",
    {},
    createJSONStorage(() => preferencesStorage)
)
```

`login.tsx`
```typescript
const [, setSession] = useAtom(sessionPreferenceAtom)
setSession({email: "email@example.com"})
```
